### ISSUE TITLE
Modifying Script with Interfaces

### ISSUE THREAD

**Issue by cpede** [oleg-shilo/cs-script#66]
I have looked at the _Modifying script without restarting_ sample. However I'm using C++/CLI as the host and uses interfaces:

```
System::Reflection::Assembly^ assscript = nullptr;
assscript = CSScript::Load(System::IO::Path::GetFullPath(scriptpath));
m_scriptI = (IMyScript^)assscript->CreateInstance("MyScript");
```

and keeping the member m_scriptI until I need to release, trying something like:

```
delete m_scriptI;
m_scriptI = nullptr;
```
However I keep getting the cache exception if I try to modify the script and load again. How do I write code in C++/CLI to be able to modify the script using interfaces?

-cpede

**Comment by oleg-shilo** [comment]
Your problem is caused by the CLR design limitation. After the assembly has been loaded it cannot be unloaded. This is a well known .NET flaw that is not going to be ever fixed. MS acknowledged it but refused to fix. 

You can read more in the "Remote vs. Local loading" section of [Script Hosting Guideline](http://www.csscript.net/help/script_hosting_guideline_.html).

There are two ways around it:
- You can use `AsmHelper` as demonstrated in _Modifying script without restarting_. The sample code that comes with NuGet package also has a sample for this one. 

    It does avoid memory leaks but it also imposes design constrains on your object model. All objects to cross the AppDomain boundaries need to be serializable or inherited from `MarshalByRefObject`.

- Alternative approach is less intrusive but still not ideal. You can load the assembly without locking the assembly file. 
    ```C#
    CSScript.GlobalSettings.InMemoryAssembly = true;
    ```
    The setting above will ensure that the assembly file can be regenerated again even if the assembly is currently loaded. The side effect of this approach is that you still do not unload the assembly but only unlock the assembly file. And since your assembly was loaded as a byte array its `Assembly.Location` property is always `null` so you cannot rely on it (if you currently do).

**Comment by cpede** [comment]
Thanks for the answer. The funny thing is that this happened when I changed my compiler from VS2008 to VS2015, and at the same time changed from your CS-Script version 2.0.0 using .NET 2 to version 2.7.4 using .NET 4? So, something changes. Before the exact same code worked without any caching problems?

But anyway, I added the call: (yes I spend like 20 minutes before I noticed the 3 times sss :-)
`CSScript::GlobalSettings->InMemoryAsssembly = true;`
and found out that I also had to set cache to false:
`CSScript::CacheEnabled = false;`
for it to work.

Thanks,
-cpede
